controller/kubernetes_controller.go

```diff
@@ -0,0 +1,389 @@
+	}
+	return pv.Spec.CSI.VolumeHandle
+}
+
+func (kc *KubernetesController) enqueuePersistentVolume(pv *v1.PersistentVolume) {
+	key, err := controller.KeyFunc(pv)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pv, err))
+		return
+	}
+	kc.queue.AddRateLimited(key)
+	return
+}
+
+func (kc *KubernetesController) enqueuePodChange(pod *v1.Pod) {
+	if _, err := controller.KeyFunc(pod); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pod, err))
+		return
+	}
+
+	for _, v := range pod.Spec.Volumes {
+		if v.VolumeSource.PersistentVolumeClaim != nil {
+			pvc, err := kc.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(v.VolumeSource.PersistentVolumeClaim.ClaimName)
+			if datastore.ErrorIsNotFound(err) {
+				continue
+			}
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pvc, err))
+				return
+			}
+			pvName := pvc.Spec.VolumeName
+			if pvName != "" {
+				kc.queue.AddRateLimited(pvName)
+			}
+		}
+	}
+	return
+}
+
+func (kc *KubernetesController) enqueuePVDeletion(pv *v1.PersistentVolume) {
+	if _, err := controller.KeyFunc(pv); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pv, err))
+		return
+	}
+	if pv.Spec.CSI != nil && pv.Spec.CSI.VolumeHandle != "" {
+		kc.pvToVolumeCache.Store(pv.Name, pv.Spec.CSI.VolumeHandle)
+	}
+	return
+}
+
+func (kc *KubernetesController) cleanupForPVDeletion(pvName string) (bool, error) {
+	volumeName, ok := kc.pvToVolumeCache.Load(pvName)
+	if !ok {
+		return false, nil
+	}
+	volume, err := kc.ds.GetVolume(volumeName.(string))
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			kc.pvToVolumeCache.Delete(pvName)
+			return true, nil
+		}
+		return false, errors.WithStack(err)
+	}
+	pv, err := kc.pvLister.Get(pvName)
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		return false, errors.WithStack(err)
+	}
+	if datastore.ErrorIsNotFound(err) || pv.DeletionTimestamp != nil {
+		ks := &volume.Status.KubernetesStatus
+		if ks.PVCName != "" && ks.LastPVCRefAt == "" {
+			volume.Status.KubernetesStatus.LastPVCRefAt = kc.nowHandler()
+		}
+		if ks.PodName != "" && ks.LastPodRefAt == "" {
+			volume.Status.KubernetesStatus.LastPodRefAt = kc.nowHandler()
+		}
+		volume.Status.KubernetesStatus.PVName = ""
+		volume.Status.KubernetesStatus.PVStatus = ""
+		volume, err = kc.ds.UpdateVolume(volume)
+		if err != nil {
+			return false, errors.WithStack(err)
+		}
+		kc.eventRecorder.Eventf(volume, v1.EventTypeNormal, EventReasonStop, "Persistent Volume %v stopped to use Longhorn volume %v", pvName, volume.Name)
+	}
+	kc.pvToVolumeCache.Delete(pvName)
+	return true, nil
+}
+
+func (kc *KubernetesController) getAssociatedPod(namespace, pvcName string) (*v1.Pod, error) {
+	pods, err := kc.pLister.Pods(namespace).List(labels.NewSelector())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list pod in namespace %s", namespace)
```

Comment:
This should be `failed to getAssociatedPod`... The listing failure will contain information like which namespace or which object.

And use `labels.Everything()` instead of `labels.NewSelector()`.